text/0004-new-pre-coordicide-consensus/0004-new-pre-coordicide-consensus.tex.md

```diff
@@ -0,0 +1,191 @@
+
+#### Algorithm
+
+Score 0 will be given if one of the following is true:
+    
+1. $t(v)<r(v)-C_1$ or $t(v)>r(v)+C'_1$
+2. $r(v) - t(otrs(v))>M$
+3. All $v_i$ satisfy $r(v)-r(v_i) > C_2$
+4. if at least one $v_i$ has a score of 0 (to enforce monotonicity)
+
+Else Score 1 will be given if exactly one $v_i$ satisfies $r(v)-r(v_i) \leq C_2$
+
+Else Score 2 will be given.
+
+#### Recommended defaults
+
+$C_1$ - 30 seconds
+
+$C'_1$ - 30 seconds
+
+$C_2$ -  2 minutes
+
+$M$ - 15 minutes
+
+### Milestone based scoring
+
+#### Configurable Values
+$C_1$ - The threshold for approving transactions with a root snapshot index that is below the latest solid one.
+
+$C_2$ - Max difference between latest solid milestone index and parent $OTRSI$.
+
+$M$ - Max difference between latest solid milestone index and current transaction $OTRSI$. Current below max depth parameter.
+
+#### Definitions
+Let $ytrsi(v)$ be `Youngest Transaction Root Snapshot Index` for transaction $v$ 
+
+Let $otrsi(v)$ be `Oldest Transaction Root Snapshot Index` for transaction $v$
+
+Let $lsmi$ be `Last solid milestone index` 
+
+#### Algorithm
+
+Score 0  will be given if one of the following is true:
+ 
+1. $lsmi-ytrsi(v)>C_1$    
+2. $lsmi-otrsi(v)>M$    
+3. both $v_i$ satisfy $lsmi-otrsi(v_i)>C_2$
+4. at least one $v_i$ has a score of 0 (to enforce monotonicity)
+
+Else Score 1 will be given if exactly one $v_i$ satisfies $lsmi -otrsi(v_i) \leq C_2$
+
+Else Score 2 will given.
+
+#### Performing the weighted random selection
+A node should have in memory the entire set of `tips` and their scores:
+
+1. Sum up the total score of the tips.
+2. Pick a random number, `r` between 1 and the sum.
+3. Traverse through the set of tips and subtract the score of each tip from `r`. Stop when you reach 0 or lower.
+
+## Compass Tip Selection
+TBD
+
+
+# Drawbacks
+
+## White Flag
+1. If we ever want to supply a proof that a `value transfer` is valid and approved we can't do so by merely supplying the path from the bundle to the approving milestone as before. A proof will require to have all the transactions that are in the past cone of the approving milestone to the genesis. However, up until now we never required to give such proofs. If we want to do easy proofs we can create merkle trees from apporved bundle hashes and add them to milestones.
+2. Everything that is `seen` is part of the tangle, including double-spend attempts. Meaning we define that possibly malicious data will be saved as part of the consensus set of the tangle.
+
+## Node Tip Selection
+1. Timestamp proposal relies on local values that differ from node to node. This can theoretically cause some tips to appear lazy to certain nodes in the network, but not lazy to others. This may lead to create competing subtangles, thus the the milestone proposal is superior. However node implementation that would like to later switch to coordicide may still consider implementing it.
```

Comment:
```suggestion
1. Timestamp proposal relies on local values that differ from node to node. This can theoretically cause some tips to appear lazy to certain nodes in the network, but not lazy to others. This may lead to competing subtangles, thus the the milestone proposal is superior. However node implementations that would like to later switch to Coordicide may still consider implementing it.
```